src/BindPattern.jl

```diff
@@ -392,6 +382,48 @@ function push_pattern!(patterns::Vector{BoundPattern}, state::BinderState, pat::
     temp
 end
 
+#
+# Infer which fields to match in a positional struct pattern by inspecting the set
+# of constructors.  It would be nice to exclude constructors that have
+# required keyword parameters, but the Julia APIs offer no simple way to determine
+# which keyword parameters have defaults.  That's becasue keyword parameters without
+# defaults are just rewritten into keyword parameters with defaults that throw an
+# exception at runtime.  So we exclude functions that have any keyword parameters.
+# If that ends up being problematic, we'll revisit the strategy.
+#
+function infer_fieldnames(type::Type, len::Int, match_positionally::Bool, location::LineNumberNode)
+    members = try
+        fieldnames(type)
+    catch ex
+        error("$(location.file):$(location.line): Could not determine the field names of `$type`.")
+    end
+
+    # If we're matching by keyword, we permit the use of any declared fields.
+    match_positionally || return members
+
+    # Search for constructor methods that have the correct number of parameters,
+    # no keyword parameters, and are not varargs.
+    meths = Method[methods(type)...]
+    meths = filter(m -> !m.isva && length(Base.kwarg_decl(m))==0, meths)
+    # drop the implicit var"#self#" argument
+    argnames = map(m -> dropfirst(Base.method_argnames(m)), meths)
+    # narrow to arg lists of the correct length where all parameter names correspond to members
+    argnames = unique(filter(l -> length(l) == len && all(n -> n in members, l), argnames))
+
+    if length(argnames) == 1
+        # found a uniquely satisfying order for member names
+        return (argnames[1]...,)
+    elseif len == length(members)
+        # no unique constructor, but the correct number of fields exist; use them
+        return members
+    elseif len > length(members)
+        error("$(location.file):$(location.line): The type `$type` has $(length(members)) fields but the pattern matches $len fields.")
```

Comment:
`matches` is maybe a bit ambiguous, suggesting that it has already matched something? Minor suggestion:

```suggestion
        error("$(location.file):$(location.line): The type `$type` has $(length(members)) fields but the pattern expects $len fields.")
```